text/0000-guesses.md

```diff
@@ -0,0 +1,258 @@
+ --> <anon>:4:12
+  |
+4 |     let t: Iter;
+  |            ^^^^ undefined or not in scope
+  |
+  = help: you can import several candidates into scope:
+  = help:   `use std::collections::binary_heap::Iter;`
+  = help:   `use std::collections::btree_map::Iter;`
+  = help:   `use std::collections::btree_set::Iter;`
+  = help:   `use std::collections::hash_map::Iter;`
+  = help:   and 8 other candidates
+```
+
+The actual span where the guess should be inserted is not shown in the command line diagnostics and has
+no influence on the display of the guess.
+
+### Special case: Only one guess
+
+In case there is only a single guess, there are multiple variants on what can happen.
+
+* replacement contains `\n`
+    * display in an extra `help`
+* No `note` attached to the main diagnostic
+    * span of the replacement is the same as the main diagnostic
+        * attach the replacement string to the guess message
+        ```
+        42 |    abc
+           |    ^^^ did you mean `abcd`
+        ```
+    * span of the replacement is part of the main diagnostic's span
+        * attach the replacement string to the guess message, even if it's technically wrong
+        ```
+        42 |    Opition<u8>
+           |    ^^^^^^^^^^^ did you mean `Option`
+    * span of the replacement contains the main diagnostic's span
+        * expand the displayed span to the entire span which should be replaced
+        ```
+        42 |    a.I
+           |    ^^^ did you mean `a::I`
+        ```
+* `note` already attched to the main diagnostic
+    * Attach the guess below the note according to the no-note rules (as if the note were the main diagnostic)
+    ```
+    5 |     ignore(|z| if false { &y } else { z });
+      |            ^^^             - `y` is borrowed here
+      |            |
+      |            may outlive borrowed value `y`
+      |            did you mean `move |z|`
+    ```
+
+# How We Teach This
+[how-we-teach-this]: #how-we-teach-this
+
+There is concern for the teachability of all the different diagnostic levels of rustc. The full current list is
+
+* fatal
+    * Bug (panic and abort)
+    * Error (abort at the next convenient stage, but continue with reporting more diagnostics)
+* Warning (print and continue)
+* sub-diagnostic
+    * Help (print in its own line and continue)
+    * Note (print inline in the textual output right after the `^^^^^` under the span)
+* code-hints
+    * Suggestion (print like a help)
+
+This list would get extended by Guesses under the code hint category
+
+The complexity is not in the diagnostic levels themselves, but in choosing the correct level for a situation. This is fairly straight-forward for errors vs warnings, but in the sub-diagnostic and code-hint levels, it is less clear.
+
+The following chart should clarify the decision process for subdiagnostics
+
+![chart](https://cloud.githubusercontent.com/assets/332036/23554529/be228b76-0025-11e7-99e7-627d60f5a328.png)
```

Comment:
The dot code is in the commit message: https://github.com/rust-lang/rfcs/pull/1941/commits/aed4d3898334c265ea76ad260a7573615f87ce38

I'll always push the dot code together with changes in the image.

You can use an online tool like http://sandbox.kidstrythisathome.com/erdos/ to generate new images and edit the graph.